src/device.rs

```diff
@@ -0,0 +1,229 @@
+                return Err(last_errno());
+            }
+
+            // SAFETY: repr(transparent)
+            let arr: NonNull<Device> = NonNull::new_unchecked(arr.cast());
+
+            let _guard = guard_on_unwind((), |()| ibv_free_device_list(arr.as_ptr().cast()));
+
+            let len: usize = num_devices.numeric_cast();
+
+            if mem::size_of::<c_int>() >= mem::size_of::<usize>() {
+                let total_size = len.saturating_mul(mem::size_of::<*mut ibv_device>());
+                assert!(total_size < usize::MAX.wrapping_div(2));
+            }
+
+            Ok(Self { arr, len })
+        }
+    }
+
+    /// Returns the slice of devices
+    #[inline]
+    #[must_use]
+    pub fn as_slice(&self) -> &[Device] {
+        // SAFETY: guaranteed by `DeviceList::available`
+        unsafe { slice::from_raw_parts(self.arr.as_ptr(), self.len) }
+    }
+}
+
+impl Drop for DeviceList {
+    #[inline]
+    fn drop(&mut self) {
+        // SAFETY: ffi
+        unsafe { ibv_free_device_list(self.ffi_ptr()) }
+    }
+}
+
+impl Deref for DeviceList {
+    type Target = [Device];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl fmt::Debug for DeviceList {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        <[Device] as fmt::Debug>::fmt(self, f)
+    }
+}
+
+impl Device {
+    /// Returns `*mut ibv_device`
+    pub(crate) fn ffi_ptr(&self) -> *mut ibv_device {
+        self.0.as_ptr()
+    }
+
+    /// Returns kernel device name
+    #[inline]
+    #[must_use]
+    pub fn c_name(&self) -> &CStr {
+        // SAFETY: ffi
+        unsafe { CStr::from_ptr(ibv_get_device_name(self.ffi_ptr())) }
+    }
+
+    /// Returns kernel device name
+    #[inline]
+    #[must_use]
+    pub fn name(&self) -> &str {
+        #[allow(clippy::expect_used)]
+        self.c_name().to_str().expect("non-utf8 device name")
+    }
+
+    /// Returns device’s node GUID
+    #[inline]
+    #[must_use]
+    pub fn guid(&self) -> Guid {
+        // SAFETY: ffi
+        unsafe { Guid(ibv_get_device_guid(self.ffi_ptr())) }
+    }
+}
+
+impl fmt::Debug for Device {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = self.name();
+        let guid = self.guid();
+        f.debug_struct("Device")
+            .field("name", &name)
+            .field("guid", &guid)
+            .finish()
+    }
+}
+
+impl Guid {
+    /// Returns the bytes of GUID in network byte order.
+    #[inline]
+    #[must_use]
+    pub fn as_bytes(&self) -> &[u8; 8] {
+        // SAFETY: transparent be64
+        unsafe { &*<*const _>::cast(self) }
+    }
+}
+
+impl fmt::Debug for Guid {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Guid({:x})", self)
+    }
+}
+
+/// Encodes `src` to a hex string and process it
+fn be64_to_hex<R>(src: __be64, case: hex_simd::AsciiCase, f: impl FnOnce(&str) -> R) -> R {
```

Comment:
Accept src as Guid is better as we can use `as_bytes` method here.

Reply:
Changed in the latest force-update.